logstash-core/lib/logstash/pipeline.rb

```diff
@@ -94,60 +112,164 @@ def filters?
+    wait_inputs
+    @logger.info("Inputs stopped")
+
+    shutdown_workers
+
+    @logger.info("Pipeline shutdown complete.")
+    @logger.terminal("Logstash shutdown completed")
+
+    # exit code
+    return 0
+  end # def run
+
+  def start_workers
+    @inflight_batches = {}
+
+    @worker_threads.clear # In case we're restarting the pipeline
     begin
       start_inputs
-      start_filters if filters?
-      start_outputs
+      @outputs.each {|o| o.register }
+      @filters.each {|f| f.register}
+
+      pipeline_workers = safe_pipeline_worker_count
+      batch_size = @settings[:pipeline_batch_size]
+      batch_delay = @settings[:pipeline_batch_delay]
+      @logger.info("Starting pipeline",
+                   :id => self.pipeline_id,
+                   :pipeline_workers => pipeline_workers,
+                   :batch_size => batch_size,
+                   :batch_delay => batch_delay)
+
+      pipeline_workers.times do |t|
+        @worker_threads << Thread.new do
+          LogStash::Util.set_thread_name("[#{pipeline_id}]>worker#{t}")
+          worker_loop(batch_size, batch_delay)
+        end
+      end
     ensure
       # it is important to garantee @ready to be true after the startup sequence has been completed
       # to potentially unblock the shutdown method which may be waiting on @ready to proceed
       @ready.make_true
     end
+  end
 
-    @logger.info("Pipeline started")
-    @logger.terminal("Logstash startup completed")
+  # Main body of what a worker threda does
+  # Repeatedly takes batches off the queu, filters, then outputs them
+  def worker_loop(batch_size, batch_delay)
+    running = true
 
-    wait_inputs
+    while running
+      # To understand the purpose behind this synchronize please read the body of take_batch
+      input_batch, shutdown_received = @input_queue_pop_mutex.synchronize { take_batch(batch_size, batch_delay) }
+      running = false if shutdown_received
+      @events_consumed.increment(input_batch.size)
+
+      filtered = filter_batch(input_batch)
+      @events_filtered.increment(filtered.size)
+
+      output_batch(filtered)
 
-    if filters?
-      shutdown_filters
-      wait_filters
-      flush_filters_to_output!(:final => true)
+      inflight_batches_synchronize { set_current_thread_inflight_batch(nil) }
     end
+  end
 
-    shutdown_outputs
-    wait_outputs
+  def take_batch(batch_size, batch_delay)
+    batch = []
+    # Since this is externally synchronized in `worker_look` wec can guarantee that the visibility of an insight batch
+    # guaranteed to be a full batch not a partial batch
+    set_current_thread_inflight_batch(batch)
+
+    shutdown_received = false
+    batch_size.times do |t|
+      event = t==0 ? @input_queue.take : @input_queue.poll(batch_delay)
+      # Exit early so each thread only gets one copy of this
+      # This is necessary to ensure proper shutdown!
+      next if event.nil?
```

Comment:
I believe the comment was meant to apply to LogStash::SHUTDOWN, not event.nil?

I'm not sure in which case event would be nil? except in the case of a misbehaving plugin (which is why we should keep it). That deserves its own comment.

I believe you're right that we can simplify the logic, but I think your example is incorrect since we should just call `next` not `break` if event is nil. We should break if Shutdown is received however, because we don't expect to keep taking events if a shutdown is Found.